### PR TITLE
Use alternative CS decomposition when it is missing from LAPACK

### PR DESCRIPTION
**Pull request by jmbr**

### CODE REVIEW

**Comment by notmgsk** [comment]
` ;     Symbol "HERMITIAN-EIG" not found in the MAGICL package.`

**Comment by jmbr** [comment]
> ` ; Symbol "HERMITIAN-EIG" not found in the MAGICL package.`

That's only available in MAGICL's master or, preferably, in https://github.com/rigetti/magicl/pull/65

**Comment by notmgsk** [comment]
> > ` ; Symbol "HERMITIAN-EIG" not found in the MAGICL package.`
> 
> That's only available in MAGICL's master or, preferably, in [rigetti/magicl#65](https://github.com/rigetti/magicl/pull/65)

Ah. Then maybe we ought to make a MAGICL release so that I will be picked up here.

**Comment by jmbr** [comment]
> Ah. Then maybe we ought to make a MAGICL release so that I will be picked up here.

Yeah, that's reasonable.

**Comment by appleby** [comment]
Would it make sense for the `csd-compat` implementation to live in the magicl package, and have magicl fallback to it if zuncsd isn't available? Or maybe just provide it separately?  I only skimmed the diff, but IIUC it looks like csd.lisp only depends on `m*` from `cl-quil`, which might also be reasonably moved into `magicl`.

**Comment by jmbr** [comment]
The reason this code presently lives in `cl-quil` is because it implements a specific case of the CSD (currently the one used in Quantum Shannon compilation). If it ever becomes a general-purpose CSD implementation, then I agree we might consider moving it to magicl.

**Comment by jmbr** [comment]
Passes the test suite. Closes #292.

**Comment by appleby** [comment]
Should we bump the dependency in `cl-quil.asd` to magicl 0.6.5?

**Comment by jmbr** [comment]
That's right.

File: src/csd.lisp [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                         (m* (magicl:conjugate-transpose vh) (m*-diag-general sigma vh)))
+                        (:hermitian-unitary
+                         (m* u (m*-diag-general sigma (magicl:conjugate-transpose u)))))))
+      (values unitary hermitian))))
+
+(defun csd-2x1 (a3 a4)
+  "Compute the 2x1 Cosine-Sine decomposition of a unitary matrix A partitioned into blocks A1 A2 as shown below:
+
+       ⎡ A3 ⎤        ⎡ A3 ⎤   ⎡ U1     ⎤ ⎡ -S ⎤
+If A = ⎢    ⎥, then  ⎢    ⎥ = ⎢        ⎥ ⎢    ⎥ V2ᴴ,
```
Comment (notmgsk) [inline src/csd.lisp]: 👍

**Comment by appleby** [comment]
> Passes the test suite. Closes #292.

By "passes the test suite" do you mean you installed a version of LAPACK without ZUNCSD and ran the test suite, or just that the new csd tests you added + rest of the test suite passes when run against an LAPACK that includes ZUNCSD.

**Comment by jmbr** [comment]
> > Passes the test suite. Closes #292.
> 
> By "passes the test suite" do you mean you installed a version of LAPACK without ZUNCSD and ran the test suite, or just that the new csd tests you added + rest of the test suite passes when run against an LAPACK that includes ZUNCSD.

The former: I used Apple's Accelerate framework (which does not include ZUNCSD) to run the entire test suite (I also double-checked that the new routines were picked up by `trace`ing them).

File: src/csd.lisp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+;;; Compute the 2x2 Cosine-Sine decomposition of an unitary matrix using a
+;;; modification of the algorithm in: E. S. Gawlik, Y. Nakatsukasa, and
+;;; B. D. Sutton, “A Backward Stable Algorithm for Computing the CS
+;;; Decomposition via the Polar Decomposition,” SIAM J. Matrix Anal. Appl.,
+;;; vol. 39, no. 3, pp. 1448–1469, Jan. 2018. DOI:10.1137/18M1182747
+
+(defun m*-diag-general (diag matrix)
+  "Returns a newly allocated matrix resulting from the product of DIAG (a diagonal real matrix) with MATRIX (a complex matrix)."
+  (declare (type magicl:matrix diag matrix)
+           (values magicl:matrix))
```
Comment (kilimanjaro) [inline src/csd.lisp]: Is this portable? If not, should we wrap it in `#+sbcl` or equivalent?

File: src/csd.lisp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            (values u1 u2 c s v2h)))))))
+
+(defun csd (matrix p q)
+  "Compute the 2x2 Cosine-Sine decomposition of MATRIX (assumed to be unitary and 2n×2n) partitioned into n×n blocks A1 A2 A3 A4 as shown below:
+
+       ⎡ A1  A3 ⎤        ⎡ A1  A3 ⎤   ⎡ U1     ⎤ ⎡ C  -S ⎤ ⎡ V1     ⎤H
+If A = ⎢        ⎥, then  ⎢        ⎥ = ⎢        ⎥ ⎢       ⎥ ⎢        ⎥,
+       ⎣ A2  A4 ⎦        ⎣ A2  A4 ⎦   ⎣     U2 ⎦ ⎣ S   C ⎦ ⎣     V2 ⎦
+
+where U1, U2, V1, and V2 are unitary and C^2 + S^2 = I. When the partition is p = q = 1, we have
```
[the remark this answers is not in the document]
Comment (kilimanjaro) [reply src/csd.lisp]: (Also applies to some of the above docstrings.)

File: tests/csd-tests.lisp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        (quil::csd a n n)
+      (let ((u1h (magicl:conjugate-transpose u1))
+            (u2h (magicl:conjugate-transpose u2))
+            (v1 (magicl:conjugate-transpose v1h))
+            (v2 (magicl:conjugate-transpose v2h)))
+        (is (m= id (quil::m* u1h u1)))
+        (is (m= id (quil::m* u2h u2)))
+        (is (m= id (quil::m* v1h v1)))
+        (is (m= id (quil::m* v2h v2)))
+        (multiple-value-bind (c s)
```
Comment (kilimanjaro) [inline tests/csd-tests.lisp]: Could use something like `(magicl:diag (mapcar #'cos theta))` to construct these.

File: src/csd.lisp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+      (values unitary hermitian))))
+
+(defun csd-2x1 (a3 a4)
+  "Compute the 2x1 Cosine-Sine decomposition of a unitary matrix A partitioned into blocks A1 A2 as shown below:
+
+       ⎡ A3 ⎤        ⎡ A3 ⎤   ⎡ U1     ⎤ ⎡ -S ⎤
+If A = ⎢    ⎥, then  ⎢    ⎥ = ⎢        ⎥ ⎢    ⎥ V2ᴴ,
+       ⎣ A4 ⎦        ⎣ A4 ⎦   ⎣     U2 ⎦ ⎣  C ⎦
+
+where U1, U2, and V2 are unitary and C^2 + S^2 = I. When the partition is p = q = 1, we have
```
Comment (kilimanjaro) [inline src/csd.lisp]: Should also note: here p,q are note even arguments.

File: src/csd.lisp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmbr) [reply src/csd.lisp]: The values of p and q determine the partitioning of the submatrices A1, A2, A3, A4. It follows the same convention as `lapack-csd` in Magicl.

File: src/csd.lisp (hunk 5 above)
Comment (jmbr) [reply src/csd.lisp]: The values of p and q here determine the sizes of A3 and A4.

The significance of the lowercase letters is to contrast complex numbers with complex (sub)matrices.

File: src/csd.lisp (hunk 2 above)
Comment (jmbr) [reply src/csd.lisp]: It compiles without errors in ECL, Clozure Common Lisp, and CLISP.

File: tests/csd-tests.lisp (hunk 4 above)
Comment (jmbr) [reply tests/csd-tests.lisp]: Good point!

File: src/csd.lisp (hunk 2 above)
Comment (kilimanjaro) [reply src/csd.lisp]: Good enough for me!

**Pull request merged**